docs/noencap-hybrid-modes.md

```diff
@@ -0,0 +1,174 @@
+
+* On AWS, the source/destination checks must be disabled on the EC2 instances of
+the Kubernetes Nodes, as described in the
+[AWS documentation](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_NAT_Instance.html#EIP_Disable_SrcDestCheck).
+
+* On Google Compute Engine, IP forwarding must be enabled on the VM instances as
+described in the [Google Cloud documentation](https://cloud.google.com/vpc/docs/using-routes#canipforward).
+
+* On Azure, there is no way to let VNet forward unknown IPs, hence Antrea
+`Hybrid` mode cannot work on Azure.
+
+If the Node network does allow Pod IPs sent out from the Nodes, you can
+configure Antrea to run in the `Hybrid` mode by setting the `trafficEncapMode`
+config option of `antrea-agent` to `hybrid`. The `trafficEncapMode` config
+option is defined in `antrea-agent.conf` of the `antrea` ConfigMap in the
+[Antrea deployment YAML](https://github.com/vmware-tanzu/antrea/blob/main/build/yamls/antrea.yml).
+
+```yaml
+  antrea-agent.conf: |
+    ... ...
+    # Determines how traffic is encapsulated. It has the following options:
+    # encap(default):    Inter-node Pod traffic is always encapsulated and Pod to external network
+    #                    traffic is SNAT'd.
+    # noEncap:           Inter-node Pod traffic is not encapsulated; Pod to external network traffic is
+    #                    SNAT'd if noSNAT is not set to true. Underlying network must be capable of
+    #                    supporting Pod traffic across IP subnets.
+    # hybrid:            noEncap if source and destination Nodes are on the same subnet, otherwise encap.
+    # networkPolicyOnly: Antrea enforces NetworkPolicy only, and utilizes CNI chaining and delegates Pod
+    #                    IPAM and connectivity to the primary CNI.
+    #
+    trafficEncapMode: hybrid
+    ... ...
+```
+
+After changing the option, you can deploy Antrea with the `Hybrid` mode by:
+
+```bash
+kubectl apply -f antrea.yml
+```
+
+## NoEncap Mode
+
+In `NoEncap` mode, Antrea never encapsulates Pod traffic. Just like `Hybrid`
+mode, the Node network needs to allow Pod IP addresses sent out from Nodes. When
+the Nodes are not in the same subnet, `NoEncap` mode additionally requires the
+Node network be able to route the Pod traffic from the source Node to the
+destination Node. There are two possibilities to enable this routing by Node
+network:
+
+* Leverage Route Controller of [Kubernetes Cloud Controller Manager](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller).
+The Kubernetes Cloud Providers that implement Route Controller can add routes
+to the cloud network routers for the Pod CIDRs of Nodes, and then the cloud
+network is able to route Pod traffic between Nodes. This Route Controller
+functionality is supported by the Cloud Provider implementations of the major
+clouds, including: [AWS](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/legacy-cloud-providers/aws),
```

Comment:
Did not find documentations for how to configure cloud providers. Probably we need more work to figure out how to configure each of them.

Reply:
For in-tree providers, as far as I know, the `--configure-cloud-routes` flag for `kube-controller-manager` needs to be set to `true`. I am not aware of any other requirement. At the moment, the flag also requires `--allocate-node-cidrs`, but that's a requirement for Antrea anyway.

I believe the flag will work for Azure, GCP, AWS. I don't know about vSphere though, maybe @dantingl knows.

Reply:
Yes, I know --configure-cloud-routes, but there can be other requirements depending on the clouds. For example, Azure seems require a separate routing table being created.

Reply:
I do not want to talk about NSX-T configuration here - there will be lots of details (T1, etc.) to talk. Hopefully Danting can enhance vSphere CP documentation for route configuration, then we can just put a link here.

Reply:
I see, there is the `--cloud-config` flag that takes an arbitrary configuration file, which is cloud-specific.